### PR TITLE
[cxx-interop] import static operator call from C++23 as member callAs…

### PR DESCRIPTION
**Pull request by hyp**
…Function functions in Swift to preserve source compatibility

### CODE REVIEW

**Comment by hyp** [comment]
This is needed to build `String.swift` in stdlib's CxxStdlib module after MSVC 17.10 update: https://github.com/microsoft/STL/commit/2fa1641fb6cd0b7d9b783debdee1923160e9b8e4#diff-6692e0333a1981f812fabb0a95f0f66c016830e2a7f3f393c990795a8c2d26da

**Comment by hyp** [comment]
@swift-ci please smoke test

**Comment by hyp** [comment]
@swift-ci please test

File: lib/ClangImporter/ImportDecl.cpp [hunk 1]
```diff
@@ -3734,6 +3734,20 @@ namespace {
     }
 
     Decl *VisitCXXMethodDecl(const clang::CXXMethodDecl *decl) {
+      // The static `operator ()` introduced in C++ 23 is still callable as an
+      // instance operator in C++, and we want to preserve the ability to call
+      // it as an instance method in Swift as well for source compatibility.
+      // Therefore, we synthesize a C++ instance member that invokes the
+      // operator and import it instead.
+      if (decl->getDeclName().getNameKind() ==
```
Comment (egorzhdan) [inline lib/ClangImporter/ImportDecl.cpp:3742]: Minor: `decl->isOverloadedOperator()`

File: test/Interop/Cxx/operators/Inputs/member-inline.h [hunk 2]
```diff
@@ -477,4 +477,14 @@ struct HasOperatorCallWithDefaultArg {
...
+class HasStaticOperatorCallBase {
+public:
+    static int operator()(int x) {
+        return x + 42;
+    }
+};
+
+class HasStaticOperatorCallDerived: public HasStaticOperatorCallBase {
+};
+
```
Comment (egorzhdan) [inline test/Interop/Cxx/operators/Inputs/member-inline.h:489]: Could you please add a test where a struct declares both static and non-static `operator()`?

File: test/Interop/Cxx/operators/member-inline.swift [hunk 3]
```diff
@@ -437,4 +437,16 @@ OperatorsTestSuite.test("HasOperatorCallWithDefaultArg.call") {
...
+  let res = h(1)
+  expectEqual(43, res)
+}
+
+OperatorsTestSuite.test("HasStaticOperatorCallDerived.call") {
+  let h = HasStaticOperatorCallDerived()
+  let res = h(0)
+  expectEqual(42, res)
+}
+
```
Comment (egorzhdan) [inline test/Interop/Cxx/operators/member-inline.swift:451]: Could you please also add a test to `test/Interop/Cxx/class/protocol-conformance-typechecker.swift` that conforms a C++ type with a static `operator()` to a Swift protocol that requires `func callAsFunction`? That would make sure that the Swift lookup table is populated correctly.

**Review by egorzhdan — COMMENTED** [review]
Thanks! I have a couple comments about tests.

File: test/Interop/Cxx/operators/Inputs/member-inline.h (hunk 2 above)
Comment (hyp) [reply test/Interop/Cxx/operators/Inputs/member-inline.h:489]: Yep! that works in C++ when the function signatures are different, so I added a test case.

**Comment by hyp** [comment]
@swift-ci please test

**Comment by hyp** [comment]
@swift-ci please test source compatibility

**Review by egorzhdan — APPROVED** [review]
LGTM except for the failing tests, thanks!

**Comment by hyp** [comment]
@swift-ci please test

**Comment by egorzhdan** [comment]
Looks like this is failing with an assertion in rebranch:
```
Assertion failed: (isInstance() && "No 'this' for static methods!"), function getThisType, file DeclCXX.cpp, line 2598.
```

https://ci.swift.org/view/Swift%20rebranch/job/oss-swift-rebranch-incremental-RA-macos/4764/

**Comment by egorzhdan** [comment]
Here's a PR with the fix for rebranch: https://github.com/swiftlang/swift/pull/75907

**Pull request merged**